core/templates/dev/head/services/contextual/UrlService.js

```diff
@@ -19,9 +19,16 @@
 
 oppia.factory('UrlService', ['$window', function($window) {
```

Comment:
In a previous comment, I wrote:

> In this and other methods, don't forget to handle encoding/decoding of URI components. Add tests for these cases as well.

I don't see any logic or tests related to that? If a URL has special chars they need to be decoded before params are returned, and similarly params with special chars need to be re-encoded before putting them back in the URL. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent for some examples.

Reply:
Oh, Ok, I thought you meant adding using addParam() and trying to get the added query back with getUrlParams(). Will try this.

Reply:
Done